iup/all_objects.py

```diff
@@ -0,0 +1,366 @@
+            self.start_date,
+            self.end_date,
+            self.interval,
+            self.init_elapsed,
+            self.init_interval,
+            self.init_incident,
+            self.init_cumulative,
+        )
+
+        setattr(self, "model", this_model)
+        return self
+
+
+class UptakeModel(abc.ABC):
+    @abc.abstractmethod
+    def fit(self, data: UptakeData) -> Self:
+        pass
+
+    @abc.abstractmethod
+    def predict(self, data: UptakeData, *args, **kwargs) -> UptakeData:
+        pass
+
+
+class LinearIncidentUptakeModel(UptakeModel):
+    def __init__(self):
+        self.model = LinearRegression()
+        pass
+
+    def fit(self, train: IncidentUptakeData) -> Self:
+        assert isinstance(train, IncidentUptakeData)
+        train.validate()
+
+        train_std = train.with_columns(
+            previous_std=(pl.col("previous") - pl.mean("previous"))
+            / pl.std("previous"),
+            elapsed_std=(pl.col("elapsed") - pl.mean("elapsed")) / pl.std("elapsed"),
+            estimate_std=(pl.col("estimate") - pl.mean("estimate"))
+            / pl.std("estimate"),
+        )
+
+        x = (
+            train_std.select(["previous_std", "elapsed_std"])
+            .with_columns(interact=pl.col("previous_std") * pl.col("elapsed_std"))
+            .to_numpy()
+        )
+        y = train_std.select(["estimate_std"]).to_numpy()
+
+        self.model.fit(x, y)
+
+        return self
+
+    def predict(
+        self,
+        train: IncidentUptakeData,
+        start_date,
+        end_date,
+        interval,
+        init_elapsed,
+        init_interval,
+        init_incident,
+        init_cumulative,
+    ) -> Self:
+        previous_mn = train["previous"].mean()
+        previous_sd = train["previous"].std()
+        elapsed_mn = train["elapsed"].mean()
+        elapsed_sd = train["elapsed"].std()
+        estimate_mn = train["estimate"].mean()
+        estimate_sd = train["estimate"].std()
+
+        incident_proj = (
+            pl.date_range(
+                start=start_date,
+                end=end_date,
+                interval=interval,
+                eager=True,
+            )
+            .alias("date")
+            .to_frame()
+            .with_columns(
+                region=pl.lit(train.select(pl.col("region").unique())),
+                elapsed=(
+                    (pl.col("date") - start_date).cast(pl.Float64)
+                    / (1000 * 60 * 60 * 24)
+                )
+                + init_elapsed,
+                season=pl.col("date").dt.year()
+                + pl.when(pl.col("date").dt.month() < 7).then(-1).otherwise(0),
+            )
+            .with_columns(
+                interval=pl.col("elapsed").diff().fill_null(init_interval),
+                season=pl.col("season").cast(pl.Utf8),
+            )
+        )
+
+        proj = np.zeros((incident_proj.shape[0]))
+        proj[0] = init_incident
+
+        for i in range(proj.shape[0] - 1):
```

Comment:
the contents of this for loop makes me queasy. What's going on here?

this should be written in a way that you can `self.model.predict(predict_x)` and get a `predict_y` and be done

although I guess you need to un-standardize your outputs? You should ask Andy about this; he and Thanasi thought about how to undo thi

Reply:
From the last known week, you predict the first unknown week. Then, from the prediction of the first unknown week, you predict the second unknown week. And so on. This is an inherently sequential calculation, so I don't know how we'll escape the need for a loop to do it. Moreover, this loop must incorporate two wrinkles that may not be obvious:

1) Before each week, you must unstandardize then restandardize the previous week's prediction, with different sets of standardization constants.
2) Although it's not currently incorporated, uncertainty should be compounded with each week's prediction. That is, it's not only last week's point estimate that determines this week's point estimate, but it's last week's full posterior distribution that determines this week's full posterior distribution.

Reply:
Oh interesting. So fundamentally the model function can only predict one week of data at a time?

I thought this was a more typical linear model, where you fit based on some points, and then you can make predictions at any number of other points. Like, if you fit a linear model to some points X1...XN, and then make prediction Y1, and then fit a new model to points {X1...XN, Y1}, then you'll get the same point estimate for a new point Y2, but with different CIs, because you now have N+1 data points as input?

Reply:
The model predicts "daily avg incident uptake during week t" based on "days elapsed since rollout at the end of week t" and "daily avg incident uptake during week t-1". Thus, a single uptake curve consisting of N report dates along with the known rollout date gives you N-1 data points for training the model.

If you want to compare the model predictions of the training data to the actual training data (i.e. if you want to evaluate the model fit), this can be done in one fell swoop with no loop, because for every outcome "daily avg incident uptake during week t" you already have the paired predictor "daily avg incident uptake during week t-1."

But if you want to make model predictions into the future, you don't have these paired predictors. The final data point in your training set serves as "daily avg incident uptake during week t-1" and you predict "daily avg incident uptake during week t" with its full posterior distribution. This prediction becomes "daily avg incident uptake during week t-1" for the next week's prediction, and so on. You are forced to move stepwise into the future. At every step, uncertainty will grow.

Reply:
We talked about this synchronously. I was thinking about models of the form $\mathrm{CU}_t = \mathrm{somefunction}(t) + \epsilon$, but you're saying something like $\mathrm{CU}_t = \sum_{t'=0}^t \mathrm{IU}_t$, where the incident uptake IU might itself be time dependent, or depend on $\mathrm{IU}_{t-1}$, or something.

So no more objections from me here